pkg/client/cache/watch_cache.go

```diff
@@ -0,0 +1,245 @@
+}
+
+func (w *WatchCache) Delete(obj interface{}) error {
+	object, resourceVersion, err := objectToVersionedRuntimeObject(obj)
+	if err != nil {
+		return err
+	}
+	event := watch.Event{watch.Deleted, object}
+
+	f := func(obj runtime.Object) error { return w.store.Delete(obj) }
+	return w.processEvent(event, resourceVersion, f)
+}
+
+func objectToVersionedRuntimeObject(obj interface{}) (runtime.Object, uint64, error) {
+	object, ok := obj.(runtime.Object)
+	if !ok {
+		return nil, 0, fmt.Errorf("obj does not implement runtime.Object interface: %v", obj)
+	}
+	meta, err := meta.Accessor(object)
+	if err != nil {
+		return nil, 0, err
+	}
+	resourceVersion, err := strconv.ParseUint(meta.ResourceVersion(), 10, 64)
+	if err != nil {
+		return nil, 0, err
+	}
+	return object, resourceVersion, nil
+}
+
+func (w *WatchCache) processEvent(event watch.Event, resourceVersion uint64, updateFunc func(runtime.Object) error) error {
+	w.Lock()
+	defer w.Unlock()
+	if w.onEvent != nil {
+		w.onEvent(event)
+	}
+	w.updateCache(resourceVersion, event)
+	w.resourceVersion = resourceVersion
+	return updateFunc(event.Object)
+}
+
+// Assumes that lock is already held for write.
+func (w *WatchCache) updateCache(resourceVersion uint64, event watch.Event) {
+	if w.endIndex == w.startIndex+w.capacity {
+		// Cache is full - remove the oldest element.
+		w.startIndex++
+	}
+	w.cache[w.endIndex%w.capacity] = watchCacheElement{resourceVersion, event}
+	w.endIndex++
+}
+
+func (w *WatchCache) List() []interface{} {
+	w.RLock()
+	defer w.RUnlock()
+	return w.store.List()
+}
+
+func (w *WatchCache) ListWithVersion() ([]interface{}, uint64) {
+	w.RLock()
+	defer w.RUnlock()
+	return w.store.List(), w.resourceVersion
+}
+
+func (w *WatchCache) ListKeys() []string {
+	w.RLock()
+	defer w.RUnlock()
+	return w.store.ListKeys()
+}
+
+func (w *WatchCache) Get(obj interface{}) (interface{}, bool, error) {
+	w.RLock()
+	defer w.RUnlock()
+	return w.store.Get(obj)
+}
+
+func (w *WatchCache) GetByKey(key string) (interface{}, bool, error) {
+	w.RLock()
+	defer w.RUnlock()
+	return w.store.GetByKey(key)
+}
+
+func (w *WatchCache) Replace(objs []interface{}) error {
+	return w.ReplaceWithVersion(objs, "0")
+}
+
+func (w *WatchCache) ReplaceWithVersion(objs []interface{}, resourceVersion string) error {
+	version, err := strconv.ParseUint(resourceVersion, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	w.Lock()
+	defer w.Unlock()
+
+	w.startIndex = 0
+	w.endIndex = 0
+	if err := w.store.Replace(objs); err != nil {
+		return err
+	}
+	w.resourceVersion = version
+	if w.onReplace != nil {
+		w.onReplace()
+	}
+	return nil
+}
+
+func (w *WatchCache) SetOnReplace(onReplace func()) {
+	w.Lock()
+	defer w.Unlock()
+	w.onReplace = onReplace
+}
+
+func (w *WatchCache) SetOnEvent(onEvent func(watch.Event)) {
+	w.Lock()
+	defer w.Unlock()
+	w.onEvent = onEvent
+}
+
+func (w *WatchCache) GetAllEventsSince(resourceVersion uint64) ([]watch.Event, error) {
```

Comment:
Late to the party, maybe done already-- can you document the behavior if you ask for e.g. 4 and 4 & 5 are in the list? Do you get both 4 and 5, or just 4? This behavior has to match etcd's behavior.